src/goto-instrument/synthesizer/synthesizer_utils.h

```diff
@@ -0,0 +1,62 @@
+/*******************************************************************\
+
+Module: Utility functions for loop invariant synthesizer.
+
+Author: Qinheping Hu
+
+\*******************************************************************/
+
+#ifndef CPROVER_GOTO_INSTRUMENT_SYNTHESIZER_SYNTHESIZER_UTILS_H
+#define CPROVER_GOTO_INSTRUMENT_SYNTHESIZER_SYNTHESIZER_UTILS_H
+
+#include <util/message.h>
+
+#include <goto-programs/goto_model.h>
+#include <goto-programs/goto_program.h>
+
+#include <goto-instrument/loop_utils.h>
+
+struct loop_idt
+{
+  irep_idt func_name;
+  unsigned int loop_number;
+
+  bool operator==(const loop_idt &o) const
+  {
+    return func_name == o.func_name && loop_number == o.loop_number;
+  }
+
+  bool operator<(const loop_idt &o) const
+  {
+    return func_name < o.func_name ||
+           (func_name == o.func_name && loop_number < o.loop_number);
+  }
+};
+
+typedef std::map<loop_idt, exprt> invariant_mapt;
```

Comment:
When we check if an invariant candidate is valid using CBMC, we need to apply it in the goto-program. However, we cannot undo the applying of loop contracts if we want to recover the goto-program and try next candidate. So we need to clone the original goto-program before applying loop contracts. If we use ```loop_head``` as a key, the loop invariant candidate cannot be correctly annotated because the ```loop_head``` does not exist in the cloned goto-program---we created a new list of instructions.